### PR TITLE
Overhauling perturbations to support custom perturbations

### PR DESCRIPTION
**Pull request by iterix**
This PR introduces Perturbations base class that can be inherited to create custom perturbations this addressing #15 

As part of this I also re-wrote the Paraphrase perturbations and fixed the error reported in #23 by @mshergad

### CODE REVIEW

File: auditor/perturbations/paraphrase.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+            model=self.model,
+            messages=payload,
+            temperature=self.temperature,
+            engine=self.engine,
+            api_version=self.api_version,
+        )
+        return Paraphrase._process_similar_sentence_reponse(response)
+
+    @staticmethod
+    def _process_similar_sentence_reponse(response):
```
Comment (iterix) [inline auditor/perturbations/paraphrase.py:86]: @mshergad This should address the bug you reported for paraphrasing single sentences. At a later point we should add a class that can paraphrase multi-line prompts.

**Comment by iterix** [comment]
@tspthomas @lior-fiddler  This PR allows users to define their own perturbations. Let me know if you have any thoughts

File: auditor/perturbations/base.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+
+class AbstractPerturbation(ABC):
+    """Abstract class to aid in creation of perturbation classes
+    """
+    def __init__(self) -> None:
+        super().__init__()
+
+    @abstractmethod
+    def perturb(self) -> List[str]:
```
Comment (lior-fiddler) [inline auditor/perturbations/base.py:12]: shouldn't the base class perturb abstract method get the prompt as a parameter ?

File: auditor/perturbations/base.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+from typing import List
+
+
+class AbstractPerturbation(ABC):
+    """Abstract class to aid in creation of perturbation classes
+    """
+    def __init__(self) -> None:
+        super().__init__()
+
+    @abstractmethod
```
Comment (lior-fiddler) [inline auditor/perturbations/base.py:11]: maybe add a comment explaining what derived concrete classes should do with this method ?

File: tests/test_perturbations.py [hunk 4]
```diff
@@ -37,4 +45,23 @@ def test_paraphrase(self):
...
+        self.num_perturbations = 4
+        self.perturber = Paraphrase(
+            num_perturbations=self.num_perturbations,
+            temperature=0.1,
+        )
+        return
+    
+    def test_paraphrase(self):
+        for prompt in TRUTHFUL_DATASET:
+            sim_prompt = self.perturber.perturb(prompt)
```
Comment (lior-fiddler) [inline tests/test_perturbations.py:61]: under the hood the test will hit openai APIs, no ? should we mock out the openai call and return a pre-canned sample response ? i.e. avoid any side effects of calling external APIs in a test

File: auditor/perturbations/base.py (hunk 2 above)
Comment (iterix) [reply auditor/perturbations/base.py]: Great point! Added, although python does not enforce the arguments when the method is over-ridden.

**Pull request merged**